mailbox/api/src/test/java/org/apache/james/mailbox/events/EventBusContract.java

```diff
@@ -0,0 +1,266 @@
+        eventBus().register(listener, new GroupA());
+        eventBus().register(listener2, new GroupB());
+
+        eventBus().dispatch(event, NO_KEYS).block();
+
+        verify(listener, times(1)).event(any());
+        verify(listener2, times(1)).event(any());
+    }
+
+    @Test
+    default void unregisteredGroupListenerShouldNotReceiveEvents() {
+        MailboxListener listener = newListener();
+        Registration registration = eventBus().register(listener, new GroupA());
+
+        registration.unregister();
+
+        eventBus().dispatch(event, NO_KEYS).block();
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void registerShouldThrowWhenAGroupIsAlreadyUsed() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+
+        eventBus().register(listener, new GroupA());
+
+        assertThatThrownBy(() -> eventBus().register(listener2, new GroupA()))
+            .isInstanceOf(GroupAlreadyRegistered.class);
+    }
+
+    @Test
+    default void registerShouldNotThrowOnAnUnregisteredGroup() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+
+        eventBus().register(listener, new GroupA()).unregister();
+
+        assertThatCode(() -> eventBus().register(listener2, new GroupA()))
+            .doesNotThrowAnyException();
+    }
+
+    @Test
+    default void dispatchShouldNotNotifyRegisteredListenerWhenEmptyKeySet() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, KEY_1);
+
+        eventBus().dispatch(event, NO_KEYS).block();
+
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void dispatchShouldNotNotifyListenerRegisteredOnOtherKeys() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, KEY_1);
+
+        eventBus().dispatch(event, ImmutableSet.of(KEY_2)).block();
+
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void dispatchShouldImpactListenerRegistered() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, KEY_1);
+
+        eventBus().dispatch(event, ImmutableSet.of(KEY_1)).block();
+
+        verify(listener, times(1)).event(any());
+    }
+
+    @Test
+    default void dispatchShouldImpactOnlyListenerRegistered() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+        eventBus().register(listener, KEY_1);
+        eventBus().register(listener2, KEY_2);
+
+        eventBus().dispatch(event, ImmutableSet.of(KEY_1)).block();
+
+        verify(listener, times(1)).event(any());
+        verifyZeroInteractions(listener2);
+    }
+
+    @Test
+    default void registerShouldAllowDuplicatedRegistration() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, KEY_1);
+        eventBus().register(listener, KEY_1);
```

Comment:
correct me if i'm wrong, but I dont see a test two listener register same key and verify the all receive event.
```java
...
        eventBus().register(listener, KEY_1);
        eventBus().register(listener2, KEY_1);
...
```

Reply:
Will add